velox/dwio/common/BitUnpacking.h

```diff
@@ -0,0 +1,663 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/dwio/common/exception/Exception.h"
+
+#if defined(__AVX2__) and defined(__BMI2__)
+
+#include <immintrin.h>
+#include <algorithm>
+#include <cstdint>
+#include <iomanip>
+#include <iostream>
+#include <optional>
+
+namespace facebook::velox::dwio::common {
+
+/// Pdep instruction masks for uint8_t. Generated by generatePdepMasks(8).
+static constexpr const uint64_t kPdepMask8[] = {
+    0x0000000000000000,
+    0x0101010101010101,
+    0x0303030303030303,
+    0x0707070707070707,
+    0x0f0f0f0f0f0f0f0f,
+    0x1f1f1f1f1f1f1f1f,
+    0x3f3f3f3f3f3f3f3f,
+    0x7f7f7f7f7f7f7f7f,
+    0xffffffffffffffff};
+
+/// Pdep instruction masks for uint16_t. Generated by generatePdepMasks(16).
+static constexpr const uint64_t kPdepMask16[] = {
+    0x0000000000000000,
+    0x0001000100010001,
+    0x0003000300030003,
+    0x0007000700070007,
+    0x000f000f000f000f,
+    0x001f001f001f001f,
+    0x003f003f003f003f,
+    0x007f007f007f007f,
+    0x00ff00ff00ff00ff,
+    0x01ff01ff01ff01ff,
+    0x03ff03ff03ff03ff,
+    0x07ff07ff07ff07ff,
+    0x0fff0fff0fff0fff,
+    0x1fff1fff1fff1fff,
+    0x3fff3fff3fff3fff,
+    0x7fff7fff7fff7fff,
+    0xffffffffffffffff};
+
+/// Pdep instruction masks for uint32_t. Generated by generatePdepMasks(32).
+static constexpr const uint64_t kPdepMask32[] = {
+    0x0000000000000000, 0x0000000100000001, 0x0000000300000003,
+    0x0000000700000007, 0x0000000f0000000f, 0x0000001f0000001f,
+    0x0000003f0000003f, 0x0000007f0000007f, 0x000000ff000000ff,
+    0x000001ff000001ff, 0x000003ff000003ff, 0x000007ff000007ff,
+    0x00000fff00000fff, 0x00001fff00001fff, 0x00003fff00003fff,
+    0x00007fff00007fff, 0x0000ffff0000ffff, 0x0001ffff0001ffff,
+    0x0003ffff0003ffff, 0x0007ffff0007ffff, 0x000fffff000fffff,
+    0x001fffff001fffff, 0x003fffff003fffff, 0x007fffff007fffff,
+    0x00ffffff00ffffff, 0x01ffffff01ffffff, 0x03ffffff03ffffff,
+    0x07ffffff07ffffff, 0x0fffffff0fffffff, 0x1fffffff1fffffff,
+    0x3fffffff3fffffff, 0x7fffffff7fffffff, 0xffffffffffffffff};
+
+// The implementation in this file uses all static inline functions instead of
+// using a class. This is to avoid function dispatch cost and facilitate
+// inlining.
```

Comment:
if it is better for moving this comment to the start of namespace?

Reply:
Moved.

Reply:
moved